### PR TITLE
Backport gh-4563 : generate dependency files

### PR DESCRIPTION
**Pull request by ev-br**
Backport of gh-4563 for 0.29.x

### CODE REVIEW

File: Cython/Build/Dependencies.py [hunk 1]
```diff
@@ -1023,6 +1027,26 @@ def copy_to_build_dir(filepath, root=os.getcwd()):
...
+                if depfile:
+                    dependencies = deps.all_dependencies(source)
+                    src_base_dir, _ = os.path.split(source)
+                    if not src_base_dir.endswith(os.sep):
+                        src_base_dir += os.sep
+                    # paths below the base_dir are relative, otherwise absolute
+                    paths = []
+                    for fname in dependencies:
+                        if (fname.startswith(src_base_dir) or 
+                            fname.startswith('.' + os.path.sep)):
```
Comment (ev-br) [inline Cython/Build/Dependencies.py:1040]: The second clause is one change w.r.t. master

File: tests/build/depfile_numpy.srctree [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+contents = [fname.replace(cy_prefix, "cy_prefix") for fname in contents]
+
+np_prefix, _ = os.path.split(np.__file__)
+contents = [fname.replace(np_prefix, "np_prefix") for fname in contents]
+
+# filter out the version number from `np_prefix/__init__.cython-30.pxd`.
+contents = [re.sub('[.]cython-[0-9]+', '', entry) for entry in contents]
+
+expected = ['cy_prefix/Includes/cpython/buffer.pxd',
+    'cy_prefix/Includes/cpython/mem.pxd',
```
Comment (ev-br) [inline tests/build/depfile_numpy.srctree:36]: 0.29.x has two extra dependencies

File: tests/build/depfile_package.srctree [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+assert sorted(contents.split()) == sorted(['test.c:', 'sub/incl.pxi', 'test.pxd', 'test.pyx']), contents
+
+
+with open("pkg/sub/test.c.dep", "r") as f:
+    contents = f.read().replace("\n", " ").replace("\\", "")
+
+contents = [os.path.relpath(entry, '.')
+            if os.path.isabs(entry) else entry for entry in contents.split()]
+assert sorted(contents) == sorted(['test.c:', 'incl.pxi', 'test.pyx', '../test.pxd']), contents
```
Comment (ev-br) [inline tests/build/depfile_package.srctree:21]: here the path to `../test.pxd` is relative

**Review by ev-br — COMMENTED** [review]
Here are the tweaks w.r.t. master, which were needed for tests to pass locally.

File: tests/build/depfile_numpy.srctree [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+contents = contents.split()
+
+cy_prefix, _ = os.path.split(Cython.__file__)
+contents = [fname.replace(cy_prefix, "cy_prefix") for fname in contents]
+
+np_prefix, _ = os.path.split(np.__file__)
+contents = [fname.replace(np_prefix, "np_prefix") for fname in contents]
+
+# filter out the version number from `np_prefix/__init__.cython-30.pxd`.
+contents = [re.sub('[.]cython-[0-9]+', '', entry) for entry in contents]
```
Comment (da-woods) [inline tests/build/depfile_numpy.srctree:33]: The `cython-30.pxd` was a Cython 3 invention. 0.29.x just ignores these files and finds the plain .pxd.

(I'm not sure if this is important here for the sake of the tests)

Comment (ev-br) [reply tests/build/depfile_numpy.srctree:33]: Do you want me to remove it? 
So far I exercised a TDD to the letter --- stopped the editor as soon as tests passed :-), to minimize the difference between patches.

Comment (da-woods) [reply tests/build/depfile_numpy.srctree:33]: Probably? I don't think these files should appear in the 0.29.x list of dependencies, so filtering them out is hiding a bug

Comment (ev-br) [reply tests/build/depfile_numpy.srctree:33]: OK, will do, after the CI gives the full list of things it does not like. One other question is the codestyle CI failure --- this does not seem related, is it known?

File: Cython/Build/Dependencies.py [hunk 5]
```diff
@@ -1023,6 +1027,26 @@ def copy_to_build_dir(filepath, root=os.getcwd()):
...
+                # write out the depfile, if requested
+                if depfile:
+                    dependencies = deps.all_dependencies(source)
+                    src_base_dir, _ = os.path.split(source)
+                    if not src_base_dir.endswith(os.sep):
+                        src_base_dir += os.sep
+                    # paths below the base_dir are relative, otherwise absolute
+                    paths = []
+                    for fname in dependencies:
+                        if (fname.startswith(src_base_dir) or
```
Comment (da-woods) [inline Cython/Build/Dependencies.py:1039]: The code-style failure looks to be trailing white-space on this line. It seems to be passing on the last commit to 0.29.x (https://github.com/cython/cython/runs/4831647955?check_suite_focus=true)

File: tests/build/depfile_numpy.srctree (hunk 4 above)
Comment (scoder) [reply tests/build/depfile_numpy.srctree]: > The `cython-30.pxd` was a Cython 3 invention. 0.29.x just ignores these files and finds the plain .pxd.

In fact, that 0.29.x ignores it is the whole point that this file suffix exists at all. :)

File: Cython/Build/Dependencies.py (hunk 5 above)
Comment (ev-br) [reply Cython/Build/Dependencies.py]: Thanks! Fixed

**Comment by ev-br** [comment]
Tried restarting the CI, but no, several jobs time out after 40 mins, and I fail to see anything obvious in the CI logs. Are timeouts expected for the 0.29.x branch?

**Comment by da-woods** [comment]
> Are timeouts expected for the 0.29.x branch?

It looks like it gets caught in some distutils/setuptools/mutliprocessing loop I think and doesn't even get past the "compile Cython" stage.

[This is the last run of the 0.29.x branch](https://github.com/cython/cython/runs/4893448052?check_suite_focus=true) and it seems to have the same problem though, so I don't think it's related to this PR

**Comment by scoder** [comment]
Thanks

**Pull request merged**